crates/ark/src/treesitter.rs

```diff
@@ -624,45 +642,154 @@ pub(crate) fn point_end_of_previous_row(
     }
 }
 
-pub(crate) struct TSQuery {
-    query: tree_sitter::Query,
+pub(crate) struct TsQuery<'ts_query> {
+    query: QueryStorage<'ts_query>,
     cursor: tree_sitter::QueryCursor,
 }
 
-impl TSQuery {
+pub(crate) enum QueryStorage<'ts_query> {
+    Owned(tree_sitter::Query),
+    Borrowed(&'ts_query tree_sitter::Query),
+}
+
+impl<'ts_query> TsQuery<'ts_query> {
+    #[allow(unused)]
     pub(crate) fn new(query_str: &str) -> anyhow::Result<Self> {
         let language = &tree_sitter_r::LANGUAGE.into();
         let query = tree_sitter::Query::new(language, query_str)
             .map_err(|err| anyhow!("Failed to compile query: {err}"))?;
 
+        Ok(Self {
+            query: QueryStorage::Owned(query),
+            cursor: tree_sitter::QueryCursor::new(),
+        })
+    }
+
+    pub(crate) fn from_query(query: &'ts_query tree_sitter::Query) -> Self {
         let cursor = tree_sitter::QueryCursor::new();
+        Self {
+            query: QueryStorage::Borrowed(query),
+            cursor,
+        }
+    }
 
-        Ok(Self { query, cursor })
+    /// Run the query on `contents` and collect all captures as (capture_name, node) pairs
+    pub(crate) fn all_captures<'tree, 'query>(
+        &'query mut self,
+        node: tree_sitter::Node<'tree>,
+        contents: &'query [u8],
```

Comment:
Do the contents not have their own `'contents` lifetime? It's tied to the `TSQuery` object?